descartes_planner/include/descartes_planner/ladder_graph.h

```diff
@@ -0,0 +1,224 @@
+#ifndef DESCARTES_LADDER_GRAPH_H
+#define DESCARTES_LADDER_GRAPH_H
+
+#include "descartes_core/trajectory_id.h"
+#include "descartes_core/trajectory_timing_constraint.h"
+
+namespace descartes_planner
+{
+
+struct Rung
+{
+  descartes_core::TrajectoryID id; // corresponds to user's input ID
+  descartes_core::TimingConstraint timing; // user input timing
+  std::vector<double> data; // joint values stored in one contiguous array
+};
+
+struct __attribute__ ((__packed__)) Edge
+{
+  double cost; // transition cost from vertex who owns this object to 'idx' in next rung
+  unsigned idx; // from THIS rung to 'idx' into the NEXT rung
+};
+
+/**
+ * @brief LadderGraph is an adjacency list based, directed graph structure with vertices
+ *        arranged into "rungs" which have connections only to vertices in the adjacent
+ *        rungs. Assumes a fixed DOF.
+ */
+class LadderGraph
+{
+public:
+  using size_type = std::size_t;
+  using EdgeList = std::vector<Edge>;
+
+  /**
+   * @brief LadderGraph
+   * @param dof The number of joints that constitute a single 'DOF'
+   */
+  explicit LadderGraph(size_type dof) noexcept
+    : dof_(dof)
+  {
+    assert(dof != 0);
+  }
+
+  /**
+   * @brief resize Resizes the internal ladder to have 'n_rung' rungs
+   * @param n_rungs Number of individual rungs
+   */
+  void resize(size_type n_rungs)
+  {
+    rungs_.resize(n_rungs);
+    edges_.resize(n_rungs);
+  }
+
+  Rung& getRung(size_type index) noexcept
+  {
+    return const_cast<Rung&>(static_cast<const LadderGraph&>(*this).getRung(index));
+  }
+
+  const Rung& getRung(size_type index) const noexcept
+  {
+    assert(index < rungs_.size());
```

Comment:
What happens if this assert fails?  Can we program this to always be successful (return an empty rung?)